nixos/modules/tasks/filesystems/nfs.nix

```diff
@@ -85,8 +85,14 @@ in
         enable = mkDefault false;
       };
 
+    systemd.services.auth-rpcgss-module =
+      {
+        unitConfig.ConditionPathExists = [ "" "/etc/krb5.keytab" ];
+      };
+
     systemd.services.rpc-gssd =
       { restartTriggers = [ nfsConfFile ];
+        unitConfig.ConditionPathExists = [ "" "/etc/krb5.keytab" ];
```

Comment:
The following works for me:

```
● rpc-gssd.service - RPC security service for NFS client and server
   Loaded: loaded (/nix/store/xi3w0cb8jrjw6kz136nljd8ayyh7zjyh-nfs-utils-2.1.1/etc/systemd/system/rpc-gssd.service; enabled; vendor preset: enabled)
  Drop-In: /nix/store/l5rpm0alkscccj92rj9iv8s6slgwl7ia-system-units/rpc-gssd.service.d
           └─overrides.conf
   Active: active (running) since Wed 2017-09-20 15:30:24 BST; 6min ago
 Main PID: 8241 (rpc.gssd)
    Tasks: 1 (limit: 4915)
   Memory: 1.0M
      CPU: 5ms
   CGroup: /system.slice/rpc-gssd.service
           └─8241 /nix/store/xi3w0cb8jrjw6kz136nljd8ayyh7zjyh-nfs-utils-2.1.1/bin/rpc.gssd
```

Reply:
I don't understand what I'm doing wrong here, because it definitely doesn't work for me after fetching this revised PR locally: https://gist.github.com/eqyiel/c77454e9bae4ef518d3fdc4e2aa88be6

What commit is your nixos based off?

I'm running off nixos-unstable of a week or so ago plus a few other unrelated changes:

```
commit 4e5f2a4d160ac0dfdf57644e583b6f34c0ff4a38 (HEAD)
Author: Ruben Maher <r@rkm.id.au>
Date:   Mon Sep 18 19:12:40 2017 +0930

    nfs-utils: set /etc/krb5.keytab as default path for rpc-gssd

    Currently the `rpc-gssd.service` has a `ConditionPathExists` clause that can
    never be met, because it's looking for stateful data inside `/nix/store`.

    `auth-rpcgss-module.service` also only starts if this file exists.

    Fixes NixOS/nixpkgs#29509.

commit 47bb274a1bb41d77af5800babbeea057c125d7df
Author: Ruben Maher <r@rkm.id.au>
Date:   Sat Sep 16 12:45:33 2017 +0930

    network-manager-applet: 1.4.6 -> 1.8.2

    Make `network-manager-applet` version match `network-manager` version.

    Ref: https://github.com/NixOS/nixpkgs/pull/29375#issuecomment-329923662

commit 4886c7b7b52792167b76adb52591df1c12ae35a8
Author: Bastian Köcher <git@kchr.de>
Date:   Thu Sep 14 15:53:39 2017 +0200

    network-manager: 1.6.2 -> 1.8.2

commit 5b91f125aa58fb0243c8e7a5f9f8c06254490f16
Author: Bastian Köcher <git@kchr.de>
Date:   Thu Sep 14 14:32:08 2017 +0200

    mailnag: 1.1.0 -> 1.2.1

commit 49446109035a8e041dcd57759af3226ab9b0337d
Author: Peter Simons <simons@cryp.to>
Date:   Thu Sep 14 14:53:47 2017 +0200

    haskell-cryptol: override happy to older version to fix the build
```

Reply:
You are just looking at the original file. The override goes to different directory.
What is the output of: `systemctl cat rpc-gssd.service` after switch?

Reply:
@Mic92 it's in that gist as well, but here:
```
eqyiel@ayanami ~/git/personal/dotfiles/lib/nixpkgs (git)-[4e5f2a4...] % systemctl cat rpc-gssd
# /nix/store/i50nrgfy858ni0lnmr2my8hfi44ybhp0-nfs-utils-2.1.1/etc/systemd/system/rpc-gssd.service
[Unit]
Description=RPC security service for NFS client and server
DefaultDependencies=no
Conflicts=umount.target
Requires=var-lib-nfs-rpc_pipefs.mount
After=var-lib-nfs-rpc_pipefs.mount

ConditionPathExists=/nix/store/i50nrgfy858ni0lnmr2my8hfi44ybhp0-nfs-utils-2.1.1/etc/krb5.keytab

PartOf=nfs-utils.service

[Service]
Type=forking
ExecStart=/nix/store/i50nrgfy858ni0lnmr2my8hfi44ybhp0-nfs-utils-2.1.1/bin/rpc.gssd

# /nix/store/g1p37g8cwy8iw5c31f24dfya1lfqblva-system-units/rpc-gssd.service.d/overrides.conf
[Unit]
ConditionPathExists=/etc/krb5.keytab
X-Restart-Triggers=/nix/store/5pgzyknq6bf809003kc5v0qcxqz7gzlg-nfs.conf

[Service]
Environment="LOCALE_ARCHIVE=/nix/store/x36zk6cx0xi71593vkk4wski2vbg3sln-glibc-locales-2.25-49/lib/locale/locale-archive"
Environment="PATH=/nix/store/v3mhywkcj5vycgclmf2j8zlpbayvxndk-coreutils-8.28/bin:/nix/store/simjpbiksmvvg2xa25v65g5hk2sr014n-findutils-4.6.0/bin:/nix/store/dlk67w96zmbh4gxi72vkxdh4pm8xdb7j-gnugrep-3.1/bin:/nix/s
Environment="TZDIR=/nix/store/ghziaza6fy9x74b9pnlc4zv8bc4cy8p2-tzdata-2016j/share/zoneinfo"
```

The problem I think is that it doesn't replace the existing `ConditionPathExists`, but appends to it.  `man systemd.unit` doesn't say _specifically_ that if you use `ConditionPathExists` mutliple times that the conditions will be AND-ed, but it does say:

> If multiple conditions are specified, the unit will be executed if all of them apply (i.e. a logical AND is applied).

Reply:
That doesn't explain why it's not failing for you though. 😬

Reply:
@Mic92 What commit are you based off?  Is it possible that there have been changes to the systemd module since then?

Reply:
My [nixpkgs fork](https://github.com/Mic92/nixpkgs/commit/ae20add94994bc575e21f5800231ba43c748470e)

my output of systemd after switch:

```console
$ systemctl cat rpc-gssd
# /nix/store/xi3w0cb8jrjw6kz136nljd8ayyh7zjyh-nfs-utils-2.1.1/etc/systemd/system/rpc-gssd.service
[Unit]
Description=RPC security service for NFS client and server
DefaultDependencies=no
Conflicts=umount.target
Requires=var-lib-nfs-rpc_pipefs.mount
After=var-lib-nfs-rpc_pipefs.mount
ConditionPathExists=/nix/store/xi3w0cb8jrjw6kz136nljd8ayyh7zjyh-nfs-utils-2.1.1/etc/krb5.keytab
PartOf=nfs-utils.service
[Service]
Type=forking
ExecStart=/nix/store/xi3w0cb8jrjw6kz136nljd8ayyh7zjyh-nfs-utils-2.1.1/bin/rpc.gssd
# /nix/store/l5rpm0alkscccj92rj9iv8s6slgwl7ia-system-units/rpc-gssd.service.d/overrides.conf
[Unit]
ConditionPathExists=
ConditionPathExists=/etc/krb5.keytab
X-Restart-Triggers=/nix/store/csyhyaim2lwn2vc5xyw6snc7dqxxg9wd-nfs.conf
[Service]
Environment="LOCALE_ARCHIVE=/nix/store/58qfin5b5l8izn31fj93rryajzwp1nqy-glibc-locales-2.25-49/lib/locale/locale-archive"
Environment="PATH=/nix/store/magv1gd9rrzlrmv7cgv2nsdpj7pqqd0f-coreutils-8.28/bin:/nix/store/3rv8bjccvg4ml7x4bq59fqwhg554hyh1-findutils-4.6.0/bin:/nix/store/bpvfrxapn3ip7fvr5
Environment="TZDIR=/nix/store/da4y54s50qcnw6wddj4pixlmdah9szxj-tzdata-2016j/share/zoneinfo"
```

Reply:
So I am on ccfce4b95831cb3ba2c67c8cb33b67b639aadefa

Reply:
I am not aware of systemd module changes since then.

Reply:
@Mic92 thanks, I'll investigate your fork as soon as I get a chance.

Reply:
Oh, @Mic92 when I built your fork and it still didn't work I realised it was because I had this line in `configuration.nix`: 

```
systemd.services.rpc-gssd.unitConfig.ConditionPathExists = lib.mkForce "/etc/krb5.keytab";
``` 

Which was clobbering your solution.

I don't understand why the first empty string resets `ConditionPathExists`, but your solution works for me:

```
eqyiel@ayanami ~ % systemctl cat rpc-gssd
# /nix/store/i50nrgfy858ni0lnmr2my8hfi44ybhp0-nfs-utils-2.1.1/etc/systemd/system/rpc-gssd.service
[Unit]
Description=RPC security service for NFS client and server
DefaultDependencies=no
Conflicts=umount.target
Requires=var-lib-nfs-rpc_pipefs.mount
After=var-lib-nfs-rpc_pipefs.mount

ConditionPathExists=/nix/store/i50nrgfy858ni0lnmr2my8hfi44ybhp0-nfs-utils-2.1.1/etc/krb5.keytab

PartOf=nfs-utils.service

[Service]
Type=forking
ExecStart=/nix/store/i50nrgfy858ni0lnmr2my8hfi44ybhp0-nfs-utils-2.1.1/bin/rpc.gssd

# /nix/store/5k85as002j1bbjb74lxvlixqr9mhbqsc-system-units/rpc-gssd.service.d/overrides.conf
[Unit]
ConditionPathExists=
ConditionPathExists=/etc/krb5.keytab
X-Restart-Triggers=/nix/store/5pgzyknq6bf809003kc5v0qcxqz7gzlg-nfs.conf

[Service]
Environment="LOCALE_ARCHIVE=/nix/store/x36zk6cx0xi71593vkk4wski2vbg3sln-glibc-locales-2.25-49/lib/locale/locale-archive"
Environment="PATH=/nix/store/v3mhywkcj5vycgclmf2j8zlpbayvxndk-coreutils-8.28/bin:/nix/store/simjpbiksmvvg2xa25v65g5hk2sr014n-findutils-4.6.0/bin:/nix/store/dlk67w96zmbh4gxi72vkxdh4pm8xdb7j-gnugrep-3.1/bin:/nix/s
Environment="TZDIR=/nix/store/ghziaza6fy9x74b9pnlc4zv8bc4cy8p2-tzdata-2016j/share/zoneinfo"




eqyiel@ayanami ~ % systemctl status rpc-gssd
● rpc-gssd.service - RPC security service for NFS client and server
   Loaded: loaded (/nix/store/i50nrgfy858ni0lnmr2my8hfi44ybhp0-nfs-utils-2.1.1/etc/systemd/system/rpc-gssd.service; enabled; vendor preset: enabled)
  Drop-In: /nix/store/5k85as002j1bbjb74lxvlixqr9mhbqsc-system-units/rpc-gssd.service.d
           └─overrides.conf
   Active: active (running) since Fri 2017-09-22 09:28:05 ACST; 6min ago
  Process: 744 ExecStart=/nix/store/i50nrgfy858ni0lnmr2my8hfi44ybhp0-nfs-utils-2.1.1/bin/rpc.gssd (code=exited, status=0/SUCCESS)
 Main PID: 750 (rpc.gssd)
    Tasks: 1 (limit: 4915)
   CGroup: /system.slice/rpc-gssd.service
           └─750 /nix/store/i50nrgfy858ni0lnmr2my8hfi44ybhp0-nfs-utils-2.1.1/bin/rpc.gssd

Sep 22 09:28:05 ayanami.maher.fyi systemd[1]: Starting RPC security service for NFS client and server...
Sep 22 09:28:05 ayanami.maher.fyi systemd[1]: Started RPC security service for NFS client and server.
```

If you have no other concerns, I would say that this ready to merge.  Thank you for your help!

Reply:
The reset on empty strings is a feature of systemd to allow overriding default values defined in other files i.e. by upstream service files.